# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [2]:
import sys
sys.path.append('../')
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [3]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [4]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [5]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [6]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [7]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [8]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [9]:
df.describe()

Season         Daynum          Wteam         Wscore   
count  145289.000000  145289.000000  145289.000000  145289.000000  \
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [10]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [11]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [12]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [13]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [14]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [15]:
# Write your code here

#Here we can see that we used nunique function to show how many unique seasons we do have.
df['Season'].nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [16]:
# Write your code here


#The team with most wins is :
most_wins_teams = df["Wteam"].value_counts().idxmax()
print('The team with most wins is:',most_wins_teams)

#The team with most lose is :
most_wins_teams = df["Wteam"].value_counts().idxmin()
print('The team with less wins is:',most_wins_teams)


#the Number of Wins for the team with most wins is : 
number_of_wins_max = df["Wteam"].value_counts().max()
print(f'The team with most wins have wins {number_of_wins_max} time.')

#the Number of Wins for the team with less wins is : 
number_of_wins_min = df["Wteam"].value_counts().min()
print(f'The team with less wins have wins {number_of_wins_min} time.')



The team with most wins is: 1181
The team with less wins is: 1327
The team with most wins have wins 819 time.
The team with less wins have wins 3 time.


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [17]:
df.iloc[[df['Wscore'].argmax()]]


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [18]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [19]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [20]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [21]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [22]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [23]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [24]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [25]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [26]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [27]:
# Write your code here
#We need to start with creating the 3 df 
season_df = df.sort_values('Season')
winning_team_df = df.sort_values('Wteam')
winning_score_Df = df.sort_values('Wscore')

#Then here we should use iloc to select the rows from index 100 to index 200 and we should keep in mind that we need it for 3 columns
Columns= ["Season", "Wteam", "Wscore"]
season_df = season_df.iloc[100:201, season_df.columns.get_indexer(Columns)]
winning_team_df = winning_team_df.iloc[100:201, winning_team_df.columns.get_indexer(Columns)]
winning_score_Df = winning_score_Df.iloc[100:201, winning_score_Df.columns.get_indexer(Columns)]

#season_df.head()
#winning_team_df.tail()
winning_score_Df.tail(30)

Season  Wteam  Wscore
6234      1986   1299      45
103943    2009   1214      45
131773    2014   1296      45
92671     2007   1277      45
139914    2015   1461      45
71318     2002   1232      45
6979      1986   1233      45
77707     2003   1410      45
134035    2014   1430      45
22029     1990   1123      45
91291     2006   1179      45
56995     1999   1200      45
9001      1987   1343      45
34997     1993   1157      45
5338      1986   1451      45
127132    2013   1458      45
6101      1986   1165      45
52406     1998   1281      45
113621    2011   1214      45
80994     2004   1104      45
110068    2010   1343      45
1601      1985   1292      45
137787    2015   1224      45
109270    2010   1396      45
123294    2012   1192      45
135433    2015   1438      45
72338     2002   1443      45
135808    2015   1185      45
129196    2013   1380      45
132791    2014   1155      45

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [28]:
#Let's start with finding the Max_Score in winning_score_Df
Max_Score = winning_score_Df.loc[winning_score_Df["Wscore"].idxmax(), "Wscore"]
print('Max_Score:',Max_Score)
# Here I'll find the Max_Season in season_df 
Max_Season = season_df.loc[season_df["Wscore"].idxmax(), "Season"]
print('Max_Season:',Max_Season)

Max_Win = winning_team_df.loc[winning_team_df["Wscore"].idxmax(), "Wteam"]
print('Max_Win:',Max_Win)


Max_Score: 45
Max_Season: 1985
Max_Win: 1102


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [29]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [30]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [31]:
# Write your code here
import numpy as nu
#Let's start with creating a new column 
df['ScoreDifference']= (df["Wscore"]-df["Lscore"])

# add a filtered_df where ScoreDifference bigger the mean ScoreDifference 
filtered_df= df[(df["ScoreDifference"]) > nu.mean(df["ScoreDifference"])]
print("The mean for Score Difference is: " , nu.mean(df["ScoreDifference"]))

filtered_df.head()

The mean for Score Difference is:  12.103311331208833


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference
0     1985      20   1228      81   1328      64    N      0               17
3     1985      25   1165      70   1432      54    H      0               16
6     1985      25   1228      64   1226      44    N      0               20
8     1985      25   1260      98   1133      80    H      0               18
10    1985      25   1307     103   1288      71    H      0               32

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [32]:
# Write your code here

# I start with finding the highest_Score_Difference
highest_Score_Difference =filtered_df["ScoreDifference"].max()

# Select the row where highest_Score_Difference
row_with_max_score_difference = filtered_df[filtered_df["ScoreDifference"] == highest_Score_Difference]

#select the needed columns
selected_columns = ["Season", "Wteam", "Lteam", "ScoreDifference"]
filtered_df_selected = row_with_max_score_difference[selected_columns]

print(filtered_df_selected)


       Season  Wteam  Lteam  ScoreDifference
44653    1996   1409   1341               91


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [33]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [34]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       N         3
       A         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [35]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [36]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [37]:
# Write your code here
df.groupby("Season")["Wscore"].mean().head()


Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
Name: Wscore, dtype: float64

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [38]:
# Write your code here
df.groupby("Wteam")["Wscore"].max()


Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [39]:
# Write your code here
Aver_Score = df.groupby(["Season", "Wteam"]).agg(AverageScore=("Wscore", "mean"))

# Finds the highest average winning score for each season
Highest_wining_score = Aver_Score.groupby("Season")["AverageScore"].idxmax()
teams_with_highest_avg_score = Aver_Score.loc[Highest_wining_score].reset_index()

teams_with_highest_avg_score

Season  Wteam  AverageScore
0     1985   1328     92.800000
1     1986   1109     91.200000
2     1987   1380     95.875000
3     1988   1258    111.750000
4     1989   1258    117.315789
5     1990   1258    126.347826
6     1991   1380    112.312500
7     1992   1380     99.642857
8     1993   1380    101.875000
9     1994   1380    106.583333
10    1995   1206    102.833333
11    1996   1206    103.600000
12    1997   1254     94.142857
13    1998   1395    101.153846
14    1999   1317     95.750000
15    2000   1395     95.647059
16    2001   1395     99.315789
17    2002   1242     92.888889
18    2003   1395     88.222222
19    2004   1194     92.142857
20    2005   1311     91.333333
21    2006   1190     89.533333
22    2007   1440     99.400000
23    2008   1440     94.500000
24    2009   1377     96.200000
25    2010   1440     95.000000
26    2011   1449     90.130435
27    2012   1370     87.400000
28    2013   1377     86.500000
29    2014   1322     92.600000
30    2015   1322     90.733333
31    2016   1146     97.142857

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [40]:
# Write your code here

wins_df = df.groupby(["Season", "Wteam"]).count().reset_index()


# Assuming "Wscore" column represents the wins
wins_df = wins_df[["Season", "Wteam", "Wscore"]]  
# Renaming the columns for clarity
wins_df.columns = ["Season", "Team", "WinsCount"] 


wins_df

Season  Team  WinsCount
0        1985  1102          5
1        1985  1103          9
2        1985  1104         21
3        1985  1106         10
4        1985  1108         19
...       ...   ...        ...
10167    2016  1460         20
10168    2016  1461         12
10169    2016  1462         27
10170    2016  1463         21
10171    2016  1464          9

[10172 rows x 3 columns]

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [41]:
# Write your code here
wins_df.loc[wins_df.groupby("Season")["WinsCount"].idxmax()]


Season  Team  WinsCount
217     1985  1385         27
342     1986  1181         32
818     1987  1424         33
861     1988  1112         31
1323    1989  1328         28
1551    1990  1247         29
1739    1991  1116         30
2088    1992  1181         28
2423    1993  1231         28
2665    1994  1163         27
2928    1995  1116         26
3357    1996  1269         31
3638    1997  1242         31
3943    1998  1242         33
4201    1999  1181         32
4709    2000  1409         29
4829    2001  1181         29
5121    2002  1153         30
5456    2003  1166         29
5981    2004  1390         29
6165    2005  1228         32
6452    2006  1181         30
6841    2007  1242         30
7209    2008  1272         33
7552    2009  1272         31
7870    2010  1242         32
8217    2011  1242         32
8565    2012  1246         32
8878    2013  1211         30
9462    2014  1455         33
9609    2015  1246         34
9955    2016  1242         29

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [42]:
#Write your code here

#here we group the df by losing team ond finde the avarge for losing score
losing_df = df.groupby(["Season","Lteam"])["Lscore"].mean().reset_index(name = "losing_score_mean")


#finding the highest avarage lowest score per. team per. Season
Highest_mean_losing_Score = losing_df.loc[losing_df.groupby("Season")["losing_score_mean"].idxmax()].reset_index()
print(Highest_mean_losing_Score)


#Highest_mean_losing_Score[Highest_mean_losing_Score["losing_score_mean"] > teams_with_highest_avg_score["AverageScore"]]

    index  Season  Lteam  losing_score_mean
0     122    1985   1260          81.000000
1     287    1986   1109          90.157895
2     820    1987   1424          88.000000
3    1074    1988   1379          89.600000
4    1270    1989   1258          99.200000
5    1562    1990   1258         117.800000
6    1854    1991   1258          95.866667
7    2149    1992   1258          86.615385
8    2436    1993   1246          85.666667
9    2870    1994   1407          91.785714
10   3007    1995   1206          87.100000
11   3476    1996   1407          88.562500
12   3643    1997   1242          94.000000
13   3958    1998   1254          89.400000
14   4366    1999   1373          84.600000
15   4699    2000   1390          79.666667
16   4837    2001   1181          83.750000
17   5285    2002   1332          82.000000
18   5673    2003   1400          81.166667
19   5742    2004   1112          81.666667
20   6246    2005   1310          82.888889
21   6426    2006   1144        

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [43]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [44]:
# Write your code here

import pandas as pd



# Create a new column 'HighScoringGame' initialized with 'No'
df['HighScoringGame'] = 'No'

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Check if the winning score is greater than 100
    if row['Wscore'] > 100:
        # Set 'HighScoringGame' to 'Yes' if the condition is met
        df.at[index, 'HighScoringGame'] = 'Yes'

df.head(12)

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference   
0     1985      20   1228      81   1328      64    N      0               17  \
1     1985      25   1106      77   1354      70    H      0                7   
2     1985      25   1112      63   1223      56    H      0                7   
3     1985      25   1165      70   1432      54    H      0               16   
4     1985      25   1192      86   1447      74    H      0               12   
5     1985      25   1218      79   1337      78    H      0                1   
6     1985      25   1228      64   1226      44    N      0               20   
7     1985      25   1242      58   1268      56    N      0                2   
8     1985      25   1260      98   1133      80    H      0               18   
9     1985      25   1305      97   1424      89    H      0                8   
10    1985      25   1307     103   1288      71    H      0               32   
11    1985      25   1344      75   1438      71    N      0                4   

   HighScoringGame  
0               No  
1               No  
2               No  
3               No  
4               No  
5               No  
6               No  
7               No  
8               No  
9               No  
10             Yes  
11              No

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [58]:
# Write your code here

# Create an empty dictionary to keep track of the game counts for each team
game_counts = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Get the team names from the 'Wteam' and 'Lteam' columns
    Wteam = row['Wteam']
    Lteam = row['Lteam']
    
    # Update the game count for Wteam
    if Wteam in game_counts:
        game_counts[Wteam] += 1
    else:
        game_counts[Wteam] = 1
    
    # Update the game count for Lteam
    if Lteam in game_counts:
        game_counts[Lteam] += 1
    else:
        game_counts[Lteam] = 1

#print(game_counts)
# add game_counts to a df 
games_df = pd.DataFrame.from_dict(game_counts, orient="index", columns=["number of games"])

#Sort the dataframe descending:
games_df.sort_values("number of games", ascending=False)

number of games
1181             1013
1314             1010
1246              999
1257              994
1242              993
...               ...
1230               64
1289               48
1446               48
1118               27
1327               22

[364 rows x 1 columns]

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [69]:
# Write your code here

# Create an empty dictionary to keep track of the highest score difference for each season
highest_score_diff = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Get the season, winning score, and losing score from the row
    season = row['Season']
    winning_score = row['Wscore']
    losing_score = row['Lscore']
    
    # Calculate the score difference
    score_diff = winning_score - losing_score
    
    # Check if the season already exists in the dictionary
    if season in highest_score_diff:
        # Update the highest score difference if the current game has a higher score difference
        if score_diff > highest_score_diff[season]:
            highest_score_diff[season] = score_diff
    else:
        # If the season does not exist in the dictionary, add it with the current score difference
        highest_score_diff[season] = score_diff

# add game_counts to a df 
highest_score_diff_df = pd.DataFrame.from_dict(highest_score_diff, orient="index", columns=["highest_score_difference"])

#Sort the dataframe descending:
highest_score_diff_df.sort_index()

highest_score_difference
1985                        60
1986                        84
1987                        73
1988                        68
1989                        70
1990                        76
1991                        68
1992                        82
1993                        81
1994                        69
1995                        74
1996                        91
1997                        80
1998                        66
1999                        75
2000                        75
2001                        75
2002                        59
2003                        72
2004                        70
2005                        64
2006                        72
2007                        74
2008                        70
2009                        68
2010                        75
2011                        69
2012                        62
2013                        58
2014                        74
2015                        69
2016                        73

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [70]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [81]:
# Write your code here

# Calculate the number of wins by each team
win_counts = df["Wteam"].value_counts()

# Calculate the number of losses by each team
loss_counts = df['Lteam'].value_counts()

#print(win_counts)
#print(loss_counts)

# Add wins and losses together to get total games played
total_games = loss_counts.add(win_counts, fill_value=0)

#sorting the total gamees
total_games_sorted =total_games.sort_values(ascending=False)

print(total_games_sorted)

1181    1013
1314    1010
1246     999
1257     994
1393     993
        ... 
1230      64
1446      48
1289      48
1118      27
1327      22
Name: count, Length: 364, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [95]:
# Write your code here

# Create a new column 'ScoreDifference'
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Group the Df by season and find the index of the row with the highest score difference for each season
idxmax= df.groupby('Season')['ScoreDifference'].idxmax()
#print(idxmax)

highest_score_diff_games = df.loc[idxmax]
highest_score_diff_games

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot   
236       1985      33   1361     128   1288      68    H      0  \
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H      0   
52600     1998      33   1395     153   1410      87    H      0   
57618     1999      58   1268     132   1317      57    H      0   
60616     2000      19   1261     112   1212      37    H      0   
65999     2001      41   1196     125   1197      50    H      0   
70737     2002      50   1328     109   1183      50    H      0   
74791     2003      33   1403     107   1311      35    H      0   
79181     2004      29   1328      94   1115      24    H      0   
83372     2005      18   1403     119   1421      55    H      0   
89542     2006      60   1400     110   1341      38    H      0   
94507     2007      59   1401     101   1212      27    H      0   
97737     2008       4   1182     129   1224      59    H      0   
103723    2009      27   1314     116   1421      48    H      0   
108405    2010      15   1397     124   1421      49    H      0   
114015    2011      22   1388     121   1152      52    H      0   
118699    2012      11   1345      96   1296      34    H      0   
124057    2013       7   1235      98   1105      40    H      0   
131373    2014      61   1256     126   1255      52    H      0   
134590    2015      11   1181     113   1342      44    H      0   
140040    2016      12   1139     144   1154      71    H      0   

        ScoreDifference HighScoringGame  
236                  60             Yes  
4731                 84             Yes  
8240                 73             Yes  
12046                68             Yes  
16677                70             Yes  
19502                76             Yes  
25161                68             Yes  
27997                82             Yes  
33858                81             Yes  
36404                69             Yes  
39858                74             Yes  
44653                91             Yes  
49033                80             Yes  
52600                66             Yes  
57618                75             Yes  
60616                75             Yes  
65999                75             Yes  
70737                59             Yes  
74791                72             Yes  
79181                70              No  
83372                64             Yes  
89542                72             Yes  
94507                74             Yes  
97737                70             Yes  
103723               68             Yes  
108405               75             Yes  
114015               69             Yes  
118699               62              No  
124057               58              No  
131373               74             Yes  
134590               69             Yes  
140040               73             Yes

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [96]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [98]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [99]:
type(df['Wscore'])

pandas.core.series.Series

In [100]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [101]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference   
0    1985      20   1228      81   1328      64    N      0               17  \
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [103]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference   
0    1985      20   1228      81   1328      64    N      0               17  \
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [104]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y